stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/messages/view/MessageListScrollHelper.kt

```diff
@@ -0,0 +1,155 @@
+                    if (hasInvisibleUnreadMessage || hasScrolledUpEnough) {
+                        scrollButtonView.setUnreadCount(unreadCount)
+                        scrollButtonView.isVisible = true
+                    } else {
+                        scrollButtonView.isVisible = false
+                    }
+                }
+            }
+        )
+    }
+
+    internal fun scrollToMessage(message: Message) {
+        recyclerView.postDelayed(
+            {
+                currentList.indexOfFirst { it is MessageListItem.MessageItem && it.message.id == message.id }
+                    .takeIf { it >= 0 }
+                    ?.let {
+                        with(recyclerView) {
+                            layoutManager
+                                ?.cast<LinearLayoutManager>()
+                                ?.scrollToPositionWithOffset(it, height / 3)
+                            post {
+                                findViewHolderForAdapterPosition(it)
+                                    ?.safeCast<BaseMessageItemViewHolder<*>>()
+                                    ?.startHighlightAnimation()
+                            }
+                        }
+                    }
+            },
+            HIGHLIGHT_MESSAGE_DELAY
+        )
+    }
+
+    internal fun onMessageListChanged(isThreadStart: Boolean, hasNewMessages: Boolean, isInitialList: Boolean) {
+        if (!hasNewMessages || adapter.currentList.isEmpty()) {
+            return
+        }
+
+        if (isThreadStart) {
+            layoutManager.scrollToPosition(0)
+        } else if (isInitialList || isLastMessageMine() || isAtBottom || alwaysScrollToBottom) {
+            layoutManager.scrollToPosition(currentList.lastIndex)
+            callback.onLastMessageRead()
+        } else {
+            refreshUnreadCount()
+            scrollButtonView.setUnreadCount(unreadCount)
+            scrollButtonView.isVisible = true
+        }
+    }
+
+    private fun refreshUnreadCount() {
+        var unreadCount = 0
+        for (i in currentList.lastIndex downTo getLastSeenItemPosition() + 1) {
+            if (currentList[i].isValid()) {
+                unreadCount++
+            }
+        }
+        this.unreadCount = unreadCount
+    }
+
+    private fun getLastSeenItemPosition(): Int {
+        val lastMessageId = if (adapter.isThread) {
+            lastSeenMessageInThread
+        } else {
+            lastSeenMessageInChannel
+        }?.getStableId()
+        return currentList.indexOfLast { it.getStableId() == lastMessageId }
+    }
+
+    private fun isLastMessageMine(): Boolean {
+        return currentList
+            .lastOrNull()
+            ?.safeCast<MessageListItem.MessageItem>()
+            ?.isMine
+            ?: false
+    }
+
+    private fun MessageListItem.isValid(): Boolean {
+        return this !is MessageListItem.TypingItem &&
+            !(this is MessageListItem.MessageItem && this.isTheirs && this.message.isDeleted())
```

Comment:
Shouldn't we be handling the other item types as well? The logic could be something like

```kotlin
this is MessageListItem.MessageItem && !(this.isTheirs && this.message.isDeleted())
```